### PR TITLE
[Xamarin.Android.Build.Tests] Reworks tests to use msbuild.

### PR DESCRIPTION
**Pull request by dellis1972**
This commit reworks the tests to work on msbuild by default.
xbuild is deprecated, so we should ensure that our system works
100% on msbuild.

So allot of changes in the system here. The hightlights are

We have to NOT use `BuildingInsideVisualStudio` when building
under `msbuild`. Setting this property to `true` has a side effect
of telling `msbuild` .. "Dont bother to build ProjectReferences".
This is because under VS that is something the IDE takes care of.
It turns out that the bug we had to do with `__NonExistantFile__`
causing the `csc` task to run ALL the time was fixed.

All of the `IsTargetSkipped` checks have been unified to call
`IsTargetSkipped` method out BuildOutput. It has also been updated
to include more target "skipped" formats which xbuild and msbuild
supports.

There is also a difference in behaviour between xbuild and msbuild.
You may well have noticed that ALL the tests that have had the

	"debug" -> "release"

changes (including the debugger attribute one) are around

	$(DebugType) == "None"

Under xbuild when $(DebugType) == "None" and $(DebugSymbols) == "True"
we end up with $(DebugType) == "None" and $(DebugSymbols) == "True".
Which results in a debug runtime and apk.

However.... under msbuild when $(DebugType) == "None"
and $(DebugSymbols) == "True" we end up with $(DebugType) == "portable"
and $(DebugSymbols) == "False".. so we get a release runtime and apk.

This is because None is NOT a valid value for $(DebugType) in msbuild,
so it resets BOTH $(DebugType) and $(DebugSymbols). So we have a
difference of behaviour between the two systems.

So in this commit we check for $(DebugType) == "None" and change it
and $(DebugSymbols) to match the msbuild behaviour. This will probably
break some QA tests.. but this is expected and the tests should be
updated to reflect this.

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Tasks/ResolveLibraryProjectImports.cs [hunk 1]
```diff
@@ -37,7 +37,10 @@ public class ResolveLibraryProjectImports : Task
 		[Required]
 		public string AssemblyIdentityMapFile { get; set; }
 
-		public string CacheFile { get; set;} 
+		public string CacheFile { get; set;}
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/ResolveLibraryProjectImports.cs]: Whitespace change?

**Comment by jonpryor** [comment]
What's troubling is that in getting the tests to "use msbuild," what the tests are asserting is changing.

...yet they're still passing, both before and after this change.

I am confused. :-/

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest.OSS.cs [hunk 2]
```diff
@@ -133,15 +133,15 @@ public partial class BuildTest : BaseTest
 				/* debugType */        "",
 				/* optimize */         true ,
 				/* embedassebmlies */  true ,
-				/* expectedResult */   "debug",
+				/* expectedResult */   "release",
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest.OSS.cs]: e.g. here we go from expecting "debug" to "release". How is this related to using msbuild? Why's this still pass under xbuild?

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/ManifestTest.OSS.cs [hunk 3]
```diff
@@ -17,7 +17,7 @@ public partial class ManifestTest : BaseTest
 			new object[] { "", true, false, },
 			new object[] { "", false, true, },
 			new object[] { "None", true, false, },
-			new object[] { "None", false, false, },
+			new object[] { "None", false, true, },
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/ManifestTest.OSS.cs]: ...and here...

Comment (jonpryor) [reply src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/ManifestTest.OSS.cs]: Guess this hits `ManifestTest.DebuggerAttribute()`:

```csharp
if (!isRelease && debugType == "None" && !builder.RunningMSBuild)
    expected = false
```

That still seems odd. Why would the presence/absence of `DebuggerAttribute` depend on xbuild vs. msbuild?

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/BuildOutput.cs [hunk 4]
```diff
@@ -49,8 +49,11 @@ public bool IsTargetSkipped (string target)
 			if (!Builder.LastBuildOutput.Contains (target))
 				throw new ArgumentException (string.Format ("Target '{0}' is not even in the build output.", target));
 			return Builder.LastBuildOutput.Contains (string.Format ("Target {0} skipped due to ", target))
-				      || Builder.LastBuildOutput.Contains (string.Format ("Skipping target \"{0}\" because its outputs are up-to-date", target))
-				      || Builder.LastBuildOutput.Contains ($"Skipping target \"{target}\" because all output files are up-to-date");
+					|| Builder.LastBuildOutput.Contains (string.Format ("Skipping target \"{0}\" because it has no outputs.", target))
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/BuildOutput.cs]: I don't think this needs to be addressed in this PR, but I have to wonder...

`Builder.LastBuildOutput` is a `string` which contains the `msbuild /v:diag` build output, which could be many *megabytes* in size, right? And this is doing possibly 5 linear searches against that output.

How does that perform?

Would...regular expressions be better?

```csharp
var toMatch = new[]{
    $"Skipping target \"{target}\" because it has no outputs.",
    $"Target \"{target}\" skipped, due to",
    $"Skipping target \"{target}\" because its outputs are up-to-date",
    $"target {target}, skipping",
    $"Skipping target \"{target}\" because all output files are up-to-date",
};
var r = new Regex (string.Join ("|", toMatch));
var m = r.Match (Builder.LastBuildOutput);
return m.Success;
```

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/Builder.cs [hunk 5]
```diff
@@ -40,8 +40,9 @@ string GetVisualStudio2017Directory ()
 		public string XABuildExe {
 			get {
 				if (IsUnix) {
-					if (!string.IsNullOrEmpty (Environment.GetEnvironmentVariable ("USE_MSBUILD"))) {
-						RunningMSBuild = true;
+					RunningMSBuild = true;
+					if (!string.IsNullOrEmpty (Environment.GetEnvironmentVariable ("USE_XBUILD"))) {
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/Builder.cs]: Shouldn't we stick with `$USE_MSBUILD` for now, and when `$USE_MSBUILD` is unset or `0` we assume `xbuild`

*Eventually* we'll make `msbuild` the default everywhere, and *then* it might -- *might* -- make sense to use `$USE_XBUILD` instead of `$USE_MSBUILD`...but *now*?

**Comment by dellis1972** [comment]
@jonpryor I was going to cover that in the commit message once this works (its broken again since the xabuild.exe updates). 

The problem is a difference in behaviour between `xbuild` and `msbuild`. You may well have noticed that ALL the tests that have had the "debug" -> "release" changes (including the debugger attribute one) are around `$(DebugType) == "None"` 

Under `xbuild` when `$(DebugType) == "None"` and `$(DebugSymbols) == "True"` we end up with `$(DebugType) == "None"` and `$(DebugSymbols) == "True"`. Which results in a `debug` runtime and apk. 

However.... under `msbuild` when `$(DebugType) == "None"` and `$(DebugSymbols) == "True"` we end up with `$(DebugType) == "portable"` and `$(DebugSymbols) == "False"`.. so we get a `release` runtime and apk. 

This is because `None` is NOT a valid value for `$(DebugType)` in msbuild, so it resets BOTH `$(DebugType)` and `$(DebugSymbols)`. So we have a difference of behaviour between the two  systems. I was thinking today we should hack our targets to force xbuild to work the same way. So we can get consistent test results without any of the hacks in the tests. But I'm not sure what the effects would be, its almost certain that some of QA's tests will fail.
